core/include/cubos/core/metrics.hpp

```diff
@@ -0,0 +1,79 @@
+/// @file
+/// @brief Metrics and profiling utilities and macros.
+/// @ingroup core
+
+#pragma once
+
+#include <chrono>
+#include <mutex>
+#include <string>
+#include <vector>
+
+namespace cubos::core
+{
+    /// @brief Represents a metric.
+    struct Metric
+    {
+        std::string mName; ///< The name of the metric.
+        double mValue;     ///< The value of the metric (as a double).
+        std::string mFile; ///< The source file where the metric was created.
+        int mLine;         ///< The line number where the metric was created.
+    };
+
+    /// @brief Singleton class that manages a collection of metrics.
+    /// @ingroup core
+    class Metrics
+    {
+    public:
+        /// @brief Deleted constructor.
+        Metrics() = delete;
+
+        /// @brief Add a metric in the pool.
+        /// @param metric The metric to add.
+        static void metric(const Metric& metric);
+
+        /// @brief Size of the metric pool.
+        /// @return Metrics count.
+        static size_t size();
+
+        /// @brief Read a metric value by its name.
+        /// @param name Metric name.
+        /// @param[out] value Buffer to store metric value.
+        /// @param offset Read count per metric.
+        /// @return Whether the metric was found.
+        static bool readValue(const std::string& name, double& value, size_t& offset);
+
+        /// @brief Search for a new unique metric name.
+        /// @param[out] name Buffer to store metric name.
+        /// @param seenCount Seen metrics count.
+        /// @return Whether a new metric was found.
+        static bool readName(std::string& name, size_t& seenCount);
+    };
+
+    /// @brief Utility to measure the execution time of a scope.
+    ///
+    /// This class measures the time between its construction and destruction,
+    /// providing a way to profile code execution time.
+    class TimeTracker
```

Comment:
ScopeProfiler perhaps would be a better name? TimeTracker seems unrelated to the metrics system